File: Data Structures & Algorithms/three-integer-sum/submission-0.cpp
class Solution {
public:
    vector<vector<int>> threeSum(vector<int>& nums) {
        vector<vector<int>> ans;

        sort(nums.begin(), nums.end());

        for (int i = 0; i < nums.size(); i++) {
            // Skip duplicates for the first number
            if (i > 0 && nums[i] == nums[i - 1]) continue;

            int target = -nums[i];
            int j = i + 1;
            int k = nums.size() - 1;

            while (j < k) {
                int sum = nums[j] + nums[k];
                if (sum == target) {
                    ans.push_back({nums[i], nums[j], nums[k]});

                    // Skip duplicates for nums[j] and nums[k]
                    while (j < k && nums[j] == nums[j + 1]) j++;
                    while (j < k && nums[k] == nums[k - 1]) k--;

                    j++;
                    k--;
                } 
                else if (sum < target) {
                    j++;
                } 
                else {
                    k--;
                }
            }
        }

        return ans;
    }
};

File: Data Structures & Algorithms/car-fleet/submission-0.ts
class Solution {
    /**
     * @param {number} target
     * @param {number[]} position
     * @param {number[]} speed
     * @return {number}
     */
    carFleet(target: number, position: number[], speed: number[]): number {
        
        const cars = position.map((p,i)=>({
            position: p,
            time: (target - p )/ speed[i]
        }))

        cars.sort((a,b) => b.position - a.position)

        let ans = 0
        let leadTime = 0

        for(let car of cars){
            if(car.time > leadTime){
                ans++
                leadTime = car.time
            }
        }   

        return ans
    }
}
